hudi-common/src/main/java/org/apache/hudi/BaseHoodieTableFileIndex.java

```diff
@@ -229,17 +238,93 @@ protected List<FileSlice> getInputFileSlices(PartitionPath partition) {
     HoodieTimeline activeTimeline = getActiveTimeline();
     Option<HoodieInstant> latestInstant = activeTimeline.lastInstant();
 
-    HoodieTableFileSystemView fileSystemView =
-        new HoodieTableFileSystemView(metaClient, activeTimeline, allFiles);
-
     Option<String> queryInstant = specifiedQueryInstant.or(() -> latestInstant.map(HoodieInstant::getTimestamp));
 
     validate(activeTimeline, queryInstant);
 
+    int parallelism = Integer.parseInt(String.valueOf(configProperties.getOrDefault(HoodieCommonConfig.TABLE_LOADING_PARALLELISM.key(),
+        HoodieCommonConfig.TABLE_LOADING_PARALLELISM.defaultValue())));
+
+    Map<PartitionPath, List<FileSlice>> cachedAllInputFileSlices;
+    long buildCacheFileSlicesLocalStart = System.currentTimeMillis();
+    if (parallelism > 0 && partitions.size() > 0) {
+
+      // convert Map<Path, FileStatus[]> to Map<String, FileStatus[]>
+      Map<String, FileStatus[]> left = partitionFilesPair.getLeft().entrySet().stream().map(entry -> {
+        String partitionPath = entry.getKey().toString();
+        FileStatus[] statuses = entry.getValue();
+        return Pair.of(partitionPath, statuses);
+      }).collect(Collectors.toMap(Pair::getKey, Pair::getValue));
+
+      Map<String, FileStatus[]> partitionFiles = combine(left, partitionFilesPair.getRight());
+
+      cachedAllInputFileSlices = buildCacheFileSlicesLocalParallel(parallelism, partitions, partitionFiles, activeTimeline, queryInstant);
+    } else {
+      FileStatus[] allFiles = combine(flatMap(partitionFilesPair.getLeft().values()), flatMap(partitionFilesPair.getRight().values()));
+      HoodieTableFileSystemView fileSystemView =
+          new HoodieTableFileSystemView(metaClient, activeTimeline, allFiles);
+
+      cachedAllInputFileSlices =  getCandidateFileSlices(partitions, queryInstant, fileSystemView);
+    }
+
+    long buildCacheFileSlicesLocalEnd = System.currentTimeMillis();
+    LOG.info(String.format("Build cache file slices, spent: %d ms", buildCacheFileSlicesLocalEnd - buildCacheFileSlicesLocalStart));
+
+    return cachedAllInputFileSlices;
+  }
+
+  private Map<PartitionPath, List<FileSlice>> buildCacheFileSlicesLocalParallel(int parallelism, List<PartitionPath> partitions, Map<String, FileStatus[]> partitionFiles,
+                                                                                HoodieTimeline activeTimeline, Option<String> queryInstant) {
+    HashMap<PartitionPath, List<FileSlice>> res = new HashMap<>();
+    parallelism = Math.max(1, Math.min(parallelism, partitionFiles.size()));
+    int totalPartitions = partitionFiles.size();
+    int cursor = 0;
+    int step = totalPartitions / parallelism;
+
+    ExecutorService pool =  Executors.newFixedThreadPool((parallelism + 1));
+    ArrayList<CompletableFuture<Map<PartitionPath, List<FileSlice>>>> futureList = new ArrayList<>(parallelism + 1);
+
+    while (cursor + step <= totalPartitions) {
```

Comment:
Can we use simple Java Streams - parallelStream here to parallelize here instead of subdividing and then using parallel streams.

Reply:
+1 for parallelStream to simplify the logic